File: Backend/services/commerce/financial-services-java/audit-service/src/main/java/com/baalvion/audit/domain/WebhookDelivery.java
package com.baalvion.audit.domain;

import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.util.UUID;

/**
 * WebhookDelivery: one attempt-tracked delivery of an event to a subscription.
 *
 * Created PENDING when an event matches a subscription; a scheduled dispatcher signs and POSTs
 * it (HMAC-SHA256), retries with exponential backoff via {@code nextAttemptAt}, and routes to
 * FAILED (dead-letter) after {@code maxAttempts}. The delivery id is the idempotency key sent
 * to the subscriber.
 */
@Entity
@Table(
  name = "webhook_deliveries",
  schema = "audit",
  indexes = {
    @Index(name = "idx_wh_del_due", columnList = "status,next_attempt_at"),
    @Index(name = "idx_wh_del_sub", columnList = "subscription_id,created_at DESC")
  }
)
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class WebhookDelivery {

  @Id
  @Column(columnDefinition = "uuid")
  private UUID id;

  @Column(columnDefinition = "uuid", nullable = false)
  private UUID tenantId;

  @Column(columnDefinition = "uuid", nullable = false)
  private UUID subscriptionId;

  @Column(length = 512, nullable = false)
  private String url;

  @Column(length = 128, nullable = false)
  private String eventType;

  @Column(columnDefinition = "jsonb", nullable = false)
  private String payload;

  @Enumerated(EnumType.STRING)
  @Column(nullable = false)
  private DeliveryStatus status;

  @Column(nullable = false)
  private int attempts;

  @Column
  private Integer responseStatus;

  @Column(columnDefinition = "TEXT")
  private String lastError;

  @Column
  private LocalDateTime nextAttemptAt;

  @Column
  private LocalDateTime deliveredAt;

  @CreationTimestamp
  @Column(nullable = false, updatable = false)
  private LocalDateTime createdAt;

  @UpdateTimestamp
  @Column(nullable = false)
  private LocalDateTime updatedAt;

  @Version
  private Long version;

  public enum DeliveryStatus {
    PENDING,
    DELIVERED,
    FAILED
  }

  @PrePersist
  protected void onCreate() {
    if (id == null) {
      id = UUID.randomUUID();
    }
    if (status == null) {
      status = DeliveryStatus.PENDING;
    }
    if (nextAttemptAt == null) {
      nextAttemptAt = LocalDateTime.now();
    }
  }
}

File: Backend/services/infrastructure/developer-service/models/webhookDelivery.js
'use strict';
module.exports = function (sequelize, DataTypes) {
    return sequelize.define('webhook_delivery', {
        id:               { type: DataTypes.UUID, primaryKey: true, defaultValue: DataTypes.UUIDV4 },
        endpoint_id:      { type: DataTypes.UUID, allowNull: false },
        org_id:           { type: DataTypes.STRING(128), allowNull: true },
        event_type:       { type: DataTypes.STRING(120), allowNull: false },
        event_id:         { type: DataTypes.UUID, allowNull: false, defaultValue: DataTypes.UUIDV4 },
        payload:          { type: DataTypes.JSONB, allowNull: false, defaultValue: {} },
        status:           { type: DataTypes.STRING(16), allowNull: false, defaultValue: 'pending' }, // pending|retrying|delivered|failed
        attempts:         { type: DataTypes.INTEGER, allowNull: false, defaultValue: 0 },
        max_attempts:     { type: DataTypes.INTEGER, allowNull: false, defaultValue: 6 },
        last_status_code: { type: DataTypes.INTEGER, allowNull: true },
        last_error:       { type: DataTypes.TEXT, allowNull: true },
        response_snippet: { type: DataTypes.TEXT, allowNull: true },
        next_attempt_at:  { type: DataTypes.DATE, allowNull: true },
        created_at:       { type: DataTypes.DATE, allowNull: false, defaultValue: DataTypes.NOW },
        delivered_at:     { type: DataTypes.DATE, allowNull: true },
    }, {
        tableName: 'webhook_deliveries', schema: 'developer', timestamps: false,
        indexes: [{ fields: ['endpoint_id'] }, { fields: ['status', 'next_attempt_at'] }, { fields: ['org_id'] }],
    });
};
